### ISSUE TITLE
yolov'4 performance is lower than yolov3 on my custom data

### ISSUE THREAD

**Issue by jonguo111** [AlexeyAB/darknet#5853]
i trained yolov3 and yolov4 on my custom data using ultralic version. however, the mAP of yolov4(mish)is lower than yolov3 and yolov4(relu). But the paper mentioned yolov4 with mish activation function has the best performance...

**Comment by AlexeyAB** [comment]
Attach all you cfg-files

**Comment by jonguo111** [comment]
Hi there,
Thank you for your reply. I just replaced the activation i.e. mish to relu and trained it from scratch. Meanwhile, I modified the filters and class numbers based on my dataset. There are 1,000 images in my dataset. All of them are 512*512. 800 for training and 200 for validation. The mAP results are presented in the following:

![yolov4-relu](https://user-images.githubusercontent.com/43916561/83883321-545ec100-a711-11ea-8878-976d0532948d.JPG)
![yolov3](https://user-images.githubusercontent.com/43916561/83883324-558fee00-a711-11ea-9655-7717ac8f9f67.JPG)
![mish](https://user-images.githubusercontent.com/43916561/83883334-57f24800-a711-11ea-86c5-fb292b068b99.JPG)
Thank you.

**Comment by AlexeyAB** [comment]
Do you mean that ReLU works better than Mish in your case?
Can you provide more information and show cfg-files?

**Comment by jonguo111** [comment]
> Do you mean that ReLU works better than Mish in your case?
> Can you provide more information and show cfg-files?
pls check the following
https://github.com/ultralytics/yolov3/blob/master/cfg/yolov4-relu.cfg

**Comment by AlexeyAB** [comment]
MISH gives +0.5 - +1% AP for the same model, if both are **Trained** and **Tested** on the same resolution, and with the same hyperparameters which can increase AP but decrease AP50 and vice versa - you must take this into account when comparing.

MISH reduces FPS less than 1%.

**Comment by AlexeyAB** [comment]
Look, MISH gives +0.9 AP : https://github.com/AlexeyAB/darknet/wiki/YOLOv4-model-zoo


Model | Size | AP
-- | -- | --
YOLOv4-Leaky | 608 | 42.9
YOLOv4-Mish | 608 | 43.8

**Comment by jonguo111** [comment]
> Look, MISH gives +0.9 AP : https://github.com/AlexeyAB/darknet/wiki/YOLOv4-model-zoo
> 
> Model	Size	AP
> YOLOv4-Leaky	608	42.9
> YOLOv4-Mish	608	43.8

Yes. But the YOLOv4-mish does not work well on my custom dataset.....ReLU may be suitable on my custom dataset

**Comment by AlexeyAB** [comment]
This is interestig:
What repository do you use for training both models?
What dataset do you use, how many images and classes?
Do you use separate training and valid datasets?
Is your dataset labeld correctly, run training with `-show_imgs` flag do you see correct bboxes?
Show screeshot of cloud of points:
`./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 608 -height 608 -show`

**Comment by jonguo111** [comment]
I used this one for training these models.
https://github.com/ultralytics/yolov3
use our custom dataset. 1000 images for training and validation. three classes.
My dataset labeled correctly and can be trained on other repo like yolact with a very high AP

**Comment by AlexeyAB** [comment]
@WongKinYiu Hi, Does https://github.com/ultralytics/yolov3 support MISH activation correctly?

**Comment by WongKinYiu** [comment]
@AlexeyAB 

Yes, It supports Mish, Swish, and LeakyReLU,
But the training of Mish need twice memory space and twice training time than Leaky.
And testing of Mish need about 20% to 30% more inference time than Leaky.

**Comment by AlexeyAB** [comment]
@WongKinYiu The most strange that MISH gives lower accuracy than Leaky for some dataset.

**Comment by WongKinYiu** [comment]
@AlexeyAB 

It is possible, you can see Mish sometimes get worse results than Leaky ReLU in Mish paper.
![image](https://user-images.githubusercontent.com/12152972/84844269-91fe0b00-b07c-11ea-876d-4bfb9b3c9964.png)

> ![yolov4-relu](https://user-images.githubusercontent.com/43916561/83883321-545ec100-a711-11ea-8878-976d0532948d.JPG)
> ![yolov3](https://user-images.githubusercontent.com/43916561/83883324-558fee00-a711-11ea-9655-7717ac8f9f67.JPG)
> ![mish](https://user-images.githubusercontent.com/43916561/83883334-57f24800-a711-11ea-86c5-fb292b068b99.JPG)

And it is also may due to the update of ultrlytics.
You can see the information in above pictures, both of yolov4-relu and yolov3 tested at batch 1, but Mish tested at batch 16.
ultrlytics update their NMS to has time-out break [v3](https://github.com/ultralytics/yolov3/blob/master/utils/utils.py#L549) [v5](https://github.com/ultralytics/yolov5/blob/master/utils/utils.py#L610). so even input same image it will gets different results in each inference, especially in large-batch inference.

It also happens in [blog](https://blog.roboflow.ai/yolov4-versus-yolov5/) and [colab](https://colab.research.google.com/drive/1nk6JDzDffS2Qs6hhMCG27pQ7HWvQPR9r) testing. Every time inference will produce difference prediction. Although it can gradually reduce inference time but it will make prediction failed in cheaper devices.
![image](https://user-images.githubusercontent.com/12152972/84843766-60387480-b07b-11ea-9274-c9c6e403152d.png)
![image](https://user-images.githubusercontent.com/12152972/84843779-69c1dc80-b07b-11ea-8fff-2bad112ce120.png)

I suggest to comment time-out break in NMS and re-test Mish model in batch 1 setting.

**Comment by laclouis5** [comment]
Same result for my small dataset, Yolo v4 gives lower AP than other versions like the one with Leaky relu. Comparison here https://github.com/AlexeyAB/darknet/issues/3874#issuecomment-549470673.

**Comment by WongKinYiu** [comment]
@laclouis5 Hello,

From your experiments, I think the problem is about anchor setting.
just change anchors of YOLOv4 to same as best model in your experiments and re-train it.

**Comment by AlexeyAB** [comment]
@laclouis5 

Your problem that you don't want to use the same anchors in both cfg-files.

> Same result for my small dataset, Yolo v4 gives lower AP than other versions like the one with Leaky relu. Comparison here #3874 (comment).

No. Author of this topic get higher accuracy for YOLOv4 (leaky) than for Yolov3 or Yolov4(mish) on his custom dataset.

**Comment by laclouis5** [comment]
@WongKinYiu Hi,

I kept the original cfg configuration except for those values that are consistent in all my experiments:
- resolution 544x544
- batch size 64
- 10,000 batches, step modified accordingly
- 6 classes, filters changed accordingly
- subdivision modified to fit on the GPU

Anchors and masks are not modified and in my cfg files they seem to be the same for Yolov4 and yolov3-optimal.

**Comment by AlexeyAB** [comment]
@laclouis5 What subdivisions did you use in both cases?

**Comment by laclouis5** [comment]
@AlexeyAB,

- [csresnext50-panet-spp-original-optimal.cfg.txt](https://github.com/AlexeyAB/darknet/files/4798139/csresnext50-panet-spp-original-optimal.cfg.txt)
- [yolov4-custom.cfg.txt](https://github.com/AlexeyAB/darknet/files/4798140/yolov4-custom.cfg.txt)

Subdivision for Yolo V4 is much higher, 64 for Yolo V4 vs 18 for optimal. I suspect that the cause is `random = 1` in Yolo V4 that requires more memory to fit higher resolutions.

**Comment by AlexeyAB** [comment]
@laclouis5 

1. Yes, I would recommend you to use resize=1.5 instead of random=1, and set subdivisions=32 or 16 (better). YOLOv4 doesn't require very high mini-batch=128 as of EfficientDet, but YOLOv4 requires at least mini-batch=4 (batch=64 subdivisions=16). So if you don't have enogh GPU-RAM, then try to set random=0 and activation=leaky, but you should set subdivisions=16

2. `csresnext50-panet-spp-original-optimal.cfg.txt` - isn't Yolov3, it was one of candidate of Yolov4. It is just a little bit slower and has a little bit lower accuracy.

**Comment by laclouis5** [comment]
@AlexeyAB, ok thanks, I will try those configurations.
